File: src/系统/03．技能系统/00．技能模板+函数/01．技能函数/03．跳跃·击飞/03．线性升降系统.lua
local ____lualib = require("lualib_bundle")
local __TS__Delete = ____lualib.__TS__Delete
local ____exports = {}
local _____53D6_5355_4F4DID, _____5C1D_8BD5_6CE8_9500Tick, _____79FB_9664_5B9E_4F8B, _____7ED3_675F_5B9E_4F8B, ____on_7EBF_6027_5347_964DTick, offTick10ms, _____6D3B_52A8_7EBF_6027_5347_964D_5217_8868, _____7EBF_6027_5347_964D_6620_5C04, _____5355_4F4D_5F53_524D_7EBF_6027_5347_964D, _____5DF2_6CE8_518CTick, ____tick_8BA1_6570
local ____00_FF0E_5171_4EAB = require("系统.03．技能系统.00．技能模板+函数.01．技能函数.03．跳跃·击飞.01．跳跃系统.00．共享")
local TICK_INTERVAL = ____00_FF0E_5171_4EAB.TICK_INTERVAL
local CENTER_TIMER_TICKS = ____00_FF0E_5171_4EAB.CENTER_TIMER_TICKS
local GetHandleId = ____00_FF0E_5171_4EAB.GetHandleId
local GetUnitFlyHeight = ____00_FF0E_5171_4EAB.GetUnitFlyHeight
local SetUnitFlyHeight = ____00_FF0E_5171_4EAB.SetUnitFlyHeight
local _____6DFB_52A0_5355_4F4D_6682_505C = ____00_FF0E_5171_4EAB["添加单位暂停"]
local _____79FB_9664_5355_4F4D_6682_505C = ____00_FF0E_5171_4EAB["移除单位暂停"]
local _____786E_4FDD_5355_4F4D_53EF_8BBE_7F6E_98DE_884C_9AD8_5EA6 = ____00_FF0E_5171_4EAB["确保单位可设置飞行高度"]
local _____5355_4F4D_5B58_6D3B = ____00_FF0E_5171_4EAB["单位存活"]
local _____9650_5236_8FDB_5EA6 = ____00_FF0E_5171_4EAB["限制进度"]
function _____53D6_5355_4F4DID(_____5355_4F4D)
    return _____5355_4F4D ~= nil and _____5355_4F4D ~= 0 and GetHandleId(_____5355_4F4D) or 0 or 0
end
function _____5C1D_8BD5_6CE8_9500Tick()
    if not _____5DF2_6CE8_518CTick or #_____6D3B_52A8_7EBF_6027_5347_964D_5217_8868 ~= 0 then
        return
    end
    _____5DF2_6CE8_518CTick = false
    ____tick_8BA1_6570 = 0
    offTick10ms(____on_7EBF_6027_5347_964DTick)
end
function _____79FB_9664_5B9E_4F8B(_____5B9E_4F8B)
    __TS__Delete(_____7EBF_6027_5347_964D_6620_5C04, _____5B9E_4F8B.id)
    if _____5355_4F4D_5F53_524D_7EBF_6027_5347_964D[_____5B9E_4F8B["单位ID"]] == _____5B9E_4F8B.id then
        __TS__Delete(_____5355_4F4D_5F53_524D_7EBF_6027_5347_964D, _____5B9E_4F8B["单位ID"])
    end
    local lastIndex = #_____6D3B_52A8_7EBF_6027_5347_964D_5217_8868 - 1
    if _____5B9E_4F8B.listIndex ~= lastIndex then
        local last = _____6D3B_52A8_7EBF_6027_5347_964D_5217_8868[lastIndex + 1]
        _____6D3B_52A8_7EBF_6027_5347_964D_5217_8868[_____5B9E_4F8B.listIndex + 1] = last
        last.listIndex = _____5B9E_4F8B.listIndex
    end
    table.remove(_____6D3B_52A8_7EBF_6027_5347_964D_5217_8868)
    _____5C1D_8BD5_6CE8_9500Tick()
end
function _____7ED3_675F_5B9E_4F8B(_____5B9E_4F8B, _____539F_56E0)
    if _____7EBF_6027_5347_964D_6620_5C04[_____5B9E_4F8B.id] ~= _____5B9E_4F8B then
        return
    end
    if _____539F_56E0 == "完成" and _____5B9E_4F8B["单位"] ~= nil and _____5B9E_4F8B["单位"] ~= 0 then
        SetUnitFlyHeight(_____5B9E_4F8B["单位"], _____5B9E_4F8B["目标高度"], 0)
    end
    if _____5B9E_4F8B["暂停单位"] then
        _____79FB_9664_5355_4F4D_6682_505C(_____5B9E_4F8B["单位"], _____5B9E_4F8B["暂停来源"])
    end
    local callback = _____5B9E_4F8B["结束回调"]
    _____79FB_9664_5B9E_4F8B(_____5B9E_4F8B)
    if callback ~= nil then
        callback(_____5B9E_4F8B["单位"], _____539F_56E0, _____5B9E_4F8B.id)
    end
end
function ____on_7EBF_6027_5347_964DTick()
    ____tick_8BA1_6570 = ____tick_8BA1_6570 + 1
    if ____tick_8BA1_6570 < CENTER_TIMER_TICKS then
        return
    end
    ____tick_8BA1_6570 = 0
    local _____672CTick_5B9E_4F8BID = {}
    do
        local i = 0
        while i < #_____6D3B_52A8_7EBF_6027_5347_964D_5217_8868 do
            _____672CTick_5B9E_4F8BID[#_____672CTick_5B9E_4F8BID + 1] = _____6D3B_52A8_7EBF_6027_5347_964D_5217_8868[i + 1].id
            i = i + 1
        end
    end
    do
        local i = 0
        while i < #_____672CTick_5B9E_4F8BID do
            do
                local _____5B9E_4F8B = _____7EBF_6027_5347_964D_6620_5C04[_____672CTick_5B9E_4F8BID[i + 1]]
                if _____5B9E_4F8B == nil then
                    goto __continue21
                end
                if not _____5355_4F4D_5B58_6D3B(_____5B9E_4F8B["单位"]) then
                    _____7ED3_675F_5B9E_4F8B(_____5B9E_4F8B, "死亡")
                    goto __continue21
                end
                if _____5B9E_4F8B["主单位死亡时中断"] and _____5B9E_4F8B["主单位"] ~= nil and _____5B9E_4F8B["主单位"] ~= 0 and not _____5355_4F4D_5B58_6D3B(_____5B9E_4F8B["主单位"]) then
                    _____7ED3_675F_5B9E_4F8B(_____5B9E_4F8B, "主单位死亡")
                    goto __continue21
                end
                _____5B9E_4F8B["已运行时间"] = _____5B9E_4F8B["已运行时间"] + TICK_INTERVAL
                local progress = _____9650_5236_8FDB_5EA6(_____5B9E_4F8B["已运行时间"] / _____5B9E_4F8B["持续时间"])
                SetUnitFlyHeight(_____5B9E_4F8B["单位"], _____5B9E_4F8B["起始高度"] + (_____5B9E_4F8B["目标高度"] - _____5B9E_4F8B["起始高度"]) * progress, 0)
                if progress >= 1 then
                    _____7ED3_675F_5B9E_4F8B(_____5B9E_4F8B, "完成")
                    goto __continue21
                end
            end
            ::__continue21::
            i = i + 1
        end
    end
end
____exports["停止线性升降"] = function(id, _____539F_56E0)
    if _____539F_56E0 == nil then
        _____539F_56E0 = "中断"
    end
    local _____5B9E_4F8B = _____7EBF_6027_5347_964D_6620_5C04[id]
    if _____5B9E_4F8B == nil then
        return false
    end
    _____7ED3_675F_5B9E_4F8B(_____5B9E_4F8B, _____539F_56E0)
    return true
end
____exports["停止单位线性升降"] = function(_____5355_4F4D, _____539F_56E0)
    if _____539F_56E0 == nil then
        _____539F_56E0 = "中断"
    end
    local id = _____5355_4F4D_5F53_524D_7EBF_6027_5347_964D[_____53D6_5355_4F4DID(_____5355_4F4D)] or 0
    local ____temp_1
    if id > 0 then
        ____temp_1 = ____exports["停止线性升降"](id, _____539F_56E0)
    else
        ____temp_1 = false
    end
    return ____temp_1
end
local ____require_result_0 = require("系统.00．核心系统.05．中心计时器")
local onTick10ms = ____require_result_0.onTick10ms
offTick10ms = ____require_result_0.offTick10ms
_____6D3B_52A8_7EBF_6027_5347_964D_5217_8868 = {}
_____7EBF_6027_5347_964D_6620_5C04 = {}
_____5355_4F4D_5F53_524D_7EBF_6027_5347_964D = {}
local _____4E0B_4E00_4E2A_7EBF_6027_5347_964DID = 0
_____5DF2_6CE8_518CTick = false
____tick_8BA1_6570 = 0
local function _____5206_914D_7EBF_6027_5347_964DID()
    _____4E0B_4E00_4E2A_7EBF_6027_5347_964DID = _____4E0B_4E00_4E2A_7EBF_6027_5347_964DID + 1
    return _____4E0B_4E00_4E2A_7EBF_6027_5347_964DID
end
local function _____6CE8_518CTick()
    if _____5DF2_6CE8_518CTick then
        return
    end
    _____5DF2_6CE8_518CTick = true
    onTick10ms(____on_7EBF_6027_5347_964DTick)
end
____exports["开始线性升降"] = function(_____5355_4F4D, _____53C2_6570)
    if not _____5355_4F4D_5B58_6D3B(_____5355_4F4D) or _____53C2_6570["持续时间"] <= 0 then
        return 0
    end
    local _____5355_4F4DID = _____53D6_5355_4F4DID(_____5355_4F4D)
    if _____5355_4F4DID <= 0 then
        return 0
    end
    ____exports["停止单位线性升降"](_____5355_4F4D, "中断")
    _____786E_4FDD_5355_4F4D_53EF_8BBE_7F6E_98DE_884C_9AD8_5EA6(_____5355_4F4D)
    local id = _____5206_914D_7EBF_6027_5347_964DID()
    local _____8D77_59CB_9AD8_5EA6 = GetUnitFlyHeight(_____5355_4F4D)
    local _____5B9E_4F8B = {
        id = id,
        listIndex = #_____6D3B_52A8_7EBF_6027_5347_964D_5217_8868,
        ["单位"] = _____5355_4F4D,
        ["单位ID"] = _____5355_4F4DID,
        ["主单位"] = _____53C2_6570["主单位"],
        ["主单位死亡时中断"] = _____53C2_6570["主单位死亡时中断"] ~= false,
        ["持续时间"] = _____53C2_6570["持续时间"],
        ["已运行时间"] = 0,
        ["起始高度"] = _____8D77_59CB_9AD8_5EA6,
        ["目标高度"] = _____8D77_59CB_9AD8_5EA6 + _____53C2_6570["高度变化"],
        ["暂停单位"] = _____53C2_6570["暂停单位"] == true,
        ["暂停来源"] = "线性升降系统:" .. tostring(id),
        ["结束回调"] = _____53C2_6570["结束回调"]
    }
    _____7EBF_6027_5347_964D_6620_5C04[id] = _____5B9E_4F8B
    _____5355_4F4D_5F53_524D_7EBF_6027_5347_964D[_____5355_4F4DID] = id
    _____6D3B_52A8_7EBF_6027_5347_964D_5217_8868[#_____6D3B_52A8_7EBF_6027_5347_964D_5217_8868 + 1] = _____5B9E_4F8B
    if _____5B9E_4F8B["暂停单位"] then
        _____6DFB_52A0_5355_4F4D_6682_505C(_____5355_4F4D, _____5B9E_4F8B["暂停来源"])
    end
    _____6CE8_518CTick()
    return id
end
____exports["单位是否正在线性升降"] = function(_____5355_4F4D)
    local id = _____5355_4F4D_5F53_524D_7EBF_6027_5347_964D[_____53D6_5355_4F4DID(_____5355_4F4D)] or 0
    return id > 0 and _____7EBF_6027_5347_964D_6620_5C04[id] ~= nil
end
return ____exports

File: TS/系统/03．技能系统/00．技能模板+函数/01．技能函数/03．跳跃·击飞/03．线性升降系统.ts
/** @noSelfInFile */

import {
  TICK_INTERVAL,
  CENTER_TIMER_TICKS,
  GetHandleId,
  GetUnitFlyHeight,
  SetUnitFlyHeight,
  添加单位暂停,
  移除单位暂停,
  确保单位可设置飞行高度,
  单位存活,
  限制进度,
} from "./01．跳跃系统/00．共享";

const { onTick10ms, offTick10ms } = require("系统.00．核心系统.05．中心计时器") as {
  onTick10ms: (this: void, callback: (this: void) => void) => void;
  offTick10ms: (this: void, callback: (this: void) => void) => void;
};

export type 线性升降结束原因 = "完成" | "中断" | "死亡" | "主单位死亡";
export type 线性升降结束回调 = (this: void, 单位: any, 原因: 线性升降结束原因, 升降ID: number) => void;

export interface 线性升降参数 {
  持续时间: number;
  /** 相对当前飞行高度的变化量；正数升空，负数下降。 */
  高度变化: number;
  暂停单位?: boolean;
  主单位?: any;
  主单位死亡时中断?: boolean;
  结束回调?: 线性升降结束回调;
}

interface 线性升降实例 {
  id: number;
  listIndex: number;
  单位: any;
  单位ID: number;
  主单位?: any;
  主单位死亡时中断: boolean;
  持续时间: number;
  已运行时间: number;
  起始高度: number;
  目标高度: number;
  暂停单位: boolean;
  暂停来源: string;
  结束回调?: 线性升降结束回调;
}

const 活动线性升降列表: 线性升降实例[] = [];
const 线性升降映射: Record<number, 线性升降实例 | undefined> = {};
const 单位当前线性升降: Record<number, number | undefined> = {};
let 下一个线性升降ID = 0;
let 已注册Tick = false;
let tick计数 = 0;

function 取单位ID(this: void, 单位: any): number {
  return (单位 != null && 单位 !== 0 ? GetHandleId(单位) : 0) || 0;
}

function 分配线性升降ID(this: void): number {
  下一个线性升降ID += 1;
  return 下一个线性升降ID;
}

function 注册Tick(this: void): void {
  if (已注册Tick) return;
  已注册Tick = true;
  onTick10ms(on线性升降Tick);
}

function 尝试注销Tick(this: void): void {
  if (!已注册Tick || 活动线性升降列表.length !== 0) return;
  已注册Tick = false;
  tick计数 = 0;
  offTick10ms(on线性升降Tick);
}

function 移除实例(this: void, 实例: 线性升降实例): void {
  delete 线性升降映射[实例.id];
  if (单位当前线性升降[实例.单位ID] === 实例.id) delete 单位当前线性升降[实例.单位ID];
  const lastIndex = 活动线性升降列表.length - 1;
  if (实例.listIndex !== lastIndex) {
    const last = 活动线性升降列表[lastIndex];
    活动线性升降列表[实例.listIndex] = last;
    last.listIndex = 实例.listIndex;
  }
  活动线性升降列表.pop();
  尝试注销Tick();
}

function 结束实例(this: void, 实例: 线性升降实例, 原因: 线性升降结束原因): void {
  if (线性升降映射[实例.id] !== 实例) return;
  if (原因 === "完成" && 实例.单位 != null && 实例.单位 !== 0) {
    SetUnitFlyHeight(实例.单位, 实例.目标高度, 0);
  }
  if (实例.暂停单位) 移除单位暂停(实例.单位, 实例.暂停来源);
  const callback = 实例.结束回调;
  移除实例(实例);
  if (callback != null) callback(实例.单位, 原因, 实例.id);
}

function on线性升降Tick(this: void): void {
  tick计数 += 1;
  if (tick计数 < CENTER_TIMER_TICKS) return;
  tick计数 = 0;
  const 本Tick实例ID: number[] = [];
  for (let i = 0; i < 活动线性升降列表.length; i++) 本Tick实例ID.push(活动线性升降列表[i].id);
  for (let i = 0; i < 本Tick实例ID.length; i++) {
    const 实例 = 线性升降映射[本Tick实例ID[i]];
    if (实例 == null) continue;
    if (!单位存活(实例.单位)) {
      结束实例(实例, "死亡");
      continue;
    }
    if (实例.主单位死亡时中断 && 实例.主单位 != null && 实例.主单位 !== 0 && !单位存活(实例.主单位)) {
      结束实例(实例, "主单位死亡");
      continue;
    }
    实例.已运行时间 += TICK_INTERVAL;
    const progress = 限制进度(实例.已运行时间 / 实例.持续时间);
    SetUnitFlyHeight(实例.单位, 实例.起始高度 + (实例.目标高度 - 实例.起始高度) * progress, 0);
    if (progress >= 1) {
      结束实例(实例, "完成");
      continue;
    }
  }
}

export function 开始线性升降(this: void, 单位: any, 参数: 线性升降参数): number {
  if (!单位存活(单位) || 参数.持续时间 <= 0) return 0;
  const 单位ID = 取单位ID(单位);
  if (单位ID <= 0) return 0;
  停止单位线性升降(单位, "中断");
  确保单位可设置飞行高度(单位);
  const id = 分配线性升降ID();
  const 起始高度 = GetUnitFlyHeight(单位);
  const 实例: 线性升降实例 = {
    id,
    listIndex: 活动线性升降列表.length,
    单位,
    单位ID,
    主单位: 参数.主单位,
    主单位死亡时中断: 参数.主单位死亡时中断 !== false,
    持续时间: 参数.持续时间,
    已运行时间: 0,
    起始高度,
    目标高度: 起始高度 + 参数.高度变化,
    暂停单位: 参数.暂停单位 === true,
    暂停来源: `线性升降系统:${id}`,
    结束回调: 参数.结束回调,
  };
  线性升降映射[id] = 实例;
  单位当前线性升降[单位ID] = id;
  活动线性升降列表.push(实例);
  if (实例.暂停单位) 添加单位暂停(单位, 实例.暂停来源);
  注册Tick();
  return id;
}

export function 停止线性升降(this: void, id: number, 原因: 线性升降结束原因 = "中断"): boolean {
  const 实例 = 线性升降映射[id];
  if (实例 == null) return false;
  结束实例(实例, 原因);
  return true;
}

export function 停止单位线性升降(this: void, 单位: any, 原因: 线性升降结束原因 = "中断"): boolean {
  const id = 单位当前线性升降[取单位ID(单位)] ?? 0;
  return id > 0 ? 停止线性升降(id, 原因) : false;
}

export function 单位是否正在线性升降(this: void, 单位: any): boolean {
  const id = 单位当前线性升降[取单位ID(单位)] ?? 0;
  return id > 0 && 线性升降映射[id] != null;
}
